src/docmanager/index.ts

```diff
@@ -0,0 +1,955 @@
+   */
+  createNew(path: string, widgetName='default', kernel?: IKernelId): Widget {
+    let widget = new Widget();
+    let manager = this._contentsManager;
+    if (widgetName === 'default') {
+      widgetName = this.listWidgetFactories(path)[0];
+    }
+    let mFactoryEx = this._getModelFactoryEx(widgetName);
+    if (!mFactoryEx) {
+      return;
+    }
+    let lang = mFactoryEx.factory.preferredLanguage(path);
+    let model = mFactoryEx.factory.createNew(lang);
+    let opts = mFactoryEx.contentsOptions;
+    if (opts.format === 'json') {
+      opts.content = model.toJSON();
+    } else {
+      opts.content = model.toString();
+    }
+    manager.save(path, opts).then(contents => {
+      let id = this._createContext(path, model, widgetName, contents);
+      this._createWidget(id, widgetName, widget, kernel);
+    });
+    return widget;
+  }
+
+  /**
+   * Get the path given a widget.
+   */
+  getPath(widget: Widget): string {
+    let id = Private.contextProperty.get(widget);
+    return this._contextManager.getPath(id);
+  }
+
+  /**
+   * Clone a widget.
+   *
+   * #### Notes
+   * This will create a new widget with the same model and context
+   * as the existing widget.
+   */
+  clone(widget: Widget): Widget {
+    let parent = new Widget();
+    let id = Private.contextProperty.get(widget);
+    let name = Private.contextProperty.get(widget);
+    this._createWidget(id, name, parent);
+    return parent;
+  }
+
+  /**
+   * Filter messages on the widget.
+   */
+  filterMessage(handler: IMessageHandler, msg: Message): boolean {
+    if (msg.type !== 'close-request') {
+      return false;
+    }
+    if (this._closeGuard) {
+      // Allow the close to propagate to the widget and its layout.
+      this._closeGuard = false;
+      return false;
+    }
+    let widget = handler as Widget;
+    let id = Private.contextProperty.get(widget);
+    let model = this._contextManager.getModel(id);
+    let context = this._contextManager.getContext(id);
+    let child = (widget.layout as PanelLayout).childAt(0);
+    let name = Private.nameProperty.get(widget);
+    // Check for a sibling widget.
+    if (!name) {
+      this._closeGuard = true;
+      widget.close();
+    }
+    let factory = this._widgetFactories[name].factory;
+    this._maybeClose(widget, model.dirty).then(result => {
+      if (!result) {
+        return result;
+      }
+      return factory.beforeClose(model, context, child);
+    }).then(result => {
+      if (result) {
+        return this._cleanupWidget(widget);
+      }
+    });
+    return true;
+  }
+
+  /**
+   * Update the path of an open document.
+   *
+   * @param oldPath - The previous path.
+   *
+   * @param newPath - The new path.
+   */
+  renameFile(oldPath: string, newPath: string): void {
+    let ids = this._contextManager.getIdsForPath(oldPath);
+    for (let id in ids) {
+      this._contextManager.rename(id, newPath);
```

Comment:
If I have two models open for a single path, do they share the same session? If they have different sessions, then this will cause the session renameNotebook command to be called twice: will it be intelligent about it? Or will there be multiple session renameNotebook calls going to the server before the first one comes back and switches all the other sessions' paths?